runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -921,20 +925,494 @@ static void VMwrtbarEvaluator(TR::Node *node, TR::Register *srcReg, TR::Register
+   // Loaded Snippet
+   if (!isInstanceField)
+      {
+      static_cast<TR::J9PPCWatchedStaticFieldSnippet *>(dataSnippet)->loadSnippet();
+      }
+   else
+      {
+      static_cast<TR::J9PPCWatchedInstanceFieldSnippet *>(dataSnippet)->loadSnippet();
+      }
+   }
+
+TR::Snippet * 
+J9::Power::TreeEvaluator::getFieldWatchInstanceSnippet(TR::CodeGenerator *cg, TR::Node *node, J9Method *m, UDATA loc, UDATA os)
+   {
+   return new (cg->trHeapMemory()) TR::J9PPCWatchedInstanceFieldSnippet(cg, node, m, loc, os);
+   }
+
+TR::Snippet * 
+J9::Power::TreeEvaluator::getFieldWatchStaticSnippet(TR::CodeGenerator *cg, TR::Node *node, J9Method *m, UDATA loc, void *fieldAddress, J9Class *fieldClass)
+   {
+   return new (cg->trHeapMemory()) TR::J9PPCWatchedStaticFieldSnippet(cg, node, m, loc, fieldAddress, fieldClass);
    }
 
 void
-J9::Power::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite, TR::Register *sideEffectRegister)
+J9::Power::TreeEvaluator::generateTestAndReportFieldWatchInstructions(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite, TR::Register *sideEffectRegister, TR::Register *valueReg, TR::Register *dataSnippetRegister)
    {
-   TR_ASSERT_FATAL(false, "This helper implements platform specific code for Fieldwatch, which is currently not supported on Power platforms.\n");
+   bool is64Bit = TR::Compiler->target.is64Bit();
+   bool isResolved = !node->getSymbolReference()->isUnresolved();
+   bool isInstanceField = node->getOpCode().isIndirect();
+   bool isAOTCompile = cg->comp()->compileRelocatableCode();
+   TR::Compilation *comp = cg->comp();
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(cg->fe());
+
+   TR::Register *scratchReg = cg->allocateRegister();
+
+   // Check if snippet has been loaded
+   if (isInstanceField && !static_cast<TR::J9PPCWatchedInstanceFieldSnippet *>(dataSnippet)->isSnippetLoaded() ||
+      (!isInstanceField && !static_cast<TR::J9PPCWatchedStaticFieldSnippet *>(dataSnippet)->isSnippetLoaded()))
+      loadFieldWatchSnippet(cg, node, dataSnippet, dataSnippetRegister, scratchReg, isInstanceField);
+
+   TR::Register *fieldClassReg = cg->allocateRegister();
+   TR::MemoryReference *classFlagsMemRef = NULL;
+   TR::Register *cndReg = cg->allocateRegister(TR_CCR);
+
+   TR::LabelSymbol* startLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol* endLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol* fieldReportLabel = generateLabelSymbol(cg);
+   startLabel->setStartInternalControlFlow();
+   endLabel->setEndInternalControlFlow();
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, startLabel);
+
+   uint8_t numOfConditions = getNumOfConditionsForReportFieldAccess(node, isResolved, isWrite,isInstanceField, cg);
+   numOfConditions++;
+   TR::RegisterDependencyConditions  *deps =  new (cg->trHeapMemory())TR::RegisterDependencyConditions(numOfConditions, numOfConditions, cg->trMemory());
```

Comment:
Are these dependencies needed?

On x86, when generating OOL code we need to add dependencies in the mainline code before branching to OOL code. This is because the x86 RA is not able to obey the dependencies in the OOL code when allocating registers in the mainline code. From what I understand, this is not a limitation on Power and Z. Thus we don't need to generate these dependencies here.

@gita-omr @fjeremic Would you be able to confirm or deny this?

Reply:
Can't speak for Powqer, but it is certainly how it works on Z. No dependencies are needed for OOL.

> This is because the x86 RA is not able to obey the dependencies in the OOL code when allocating registers in the mainline code.

I think we should open an issue if this is indeed the case. Sounds like a problem we should tackle as on Z creating OOLs and using them is trivial without any bookkeeping.

Reply:
Sorry, I am not following. Is the question whether the deps are needed at all or where they should be generated?

Reply:
The question is do we need to add dependencies when branching to OOL code?

I thought X was the only one that is different in that we do need to specify the dependencies (due to limitations in their register allocator).

Reply:
To further explain the complexity here. We first create these dependencies here then add `sideEffectRegister` to the condition list on line 1336 below. Following this, we invoke `generateReportFieldAccessOutlinedInstructions` on line 1342 and pass this `deps` reference.

Once inside `generateReportFieldAccessOutlinedInstructions` we further add to these dependencies (starting on line 1136). Finally, we attach these dependencies to the call instruction that is generated on line 1160.

I understand that we need to add dependencies when making a call. But my question is if we need to attach the `sideEffectRegister` as a dependency because this is in mainline code (i.e. is the first dependency from mainline code needed on line 1336). Here's an example of this function on Z. We don't add any dependencies in the mainline code:
https://github.com/eclipse/openj9/blob/2de5d675169cc99634d4f46db3bad7e8f9283023/runtime/compiler/z/codegen/J9TreeEvaluator.cpp#L4019-L4087

Reply:
Additionally, as this code currently stands, we will add the `sideEffectRegister` dependency twice for instance fields. We add it the first time below on line 1336, and then again inside `generateReportFieldAccessOutlinedInstructions` on line 1140. I don't think that's correct.

Reply:
See https://github.com/eclipse/openj9/pull/6234/commits/59478ec9e611a0f51d70efd62ff92055afca576a

Reply:
To answer your question. Depen

> Additionally, as this code currently stands, we will add the `sideEffectRegister` dependency twice for instance fields. We add it the first time below on line 1336, and then again inside `generateReportFieldAccessOutlinedInstructions` on line 1140. I don't think that's correct.

See https://github.com/eclipse/openj9/pull/6234/commits/59478ec9e611a0f51d70efd62ff92055afca576a

Reply:
To answer your question. Any register used in OOL code would need to be added to the dependencies on Power. When register demand or pressure is high there is a risk of inconsistent spill or refill over IL not aware branches if these registers are not added to the dependency. The dependency allows us to reserve the appropriate registers.

Reply:
As far as I can see `deps` is only used within `generateReportFieldAccessOutlinedInstructions`, so we definitely don't need to pass dependencies as an argument. Eliminating this parameter will standardize the function definition of `generateReportFieldAccessOutlinedInstructions` across  platforms. Can we make this change?

Reply:
Additionally, we no longer add `sideEffectRegister` as a dependency unconditionally here. So if the explanation here is correct: https://github.com/eclipse/openj9/pull/6234#discussion_r297242299

then wouldn't this be a bug? As we make use of `sideEffectRegister` in main line code but don't add it to the dependencies in all scenarios (this is what we do on X.. and from what I can understand the comment above suggests that we need to do something similar on Power).

Reply:
@dchopra001 Yes. It was removed as a unconditional dependency in https://github.com/eclipse/openj9/commit/59478ec9e611a0f51d70efd62ff92055afca576a

Reply:
@AlenBadel I think you're misunderstanding my question. My question is whether we should add it back as an unconditional dependency. The reason being that except for when we have a static resolved field, we use this register in mainline code. However, after branching to OOL code we only add this `sideEffectRegister` to the dependency for instance fields. 

If your comment (https://github.com/eclipse/openj9/pull/6234#discussion_r297242299) is accurate, then don't we need to add `sideEffectRegister` to the dependency list for all cases where it was used in the mainline code?

Reply:
Right. To my knowledge we only need to add  `sideEffectRegister` as a dependency in the cases it's used in OOL. The decision is not dependent on if it was used in the mainline code.

Reply:
Okay so in that case power is more similar to Z than it is to X instead.

Reply:
@fjeremic
Moved deps within the method. 
See https://github.com/eclipse/openj9/pull/6234/commits/a5abb8d230c501f681024349304c2305379d94e5